src/Microsoft.AspNetCore.Testing/Functional/ProjectToolScenario.cs

```diff
@@ -0,0 +1,198 @@
+            if (string.IsNullOrEmpty(versionSuffix))
+            {
+               versionSuffix = "*";
+            }
+
+            var projectJson = JObject.Parse(File.ReadAllText(projectFile));
+            projectJson.Add("tools",
+              new JObject(
+                new JProperty(toolName,
+                  new JObject(
+                    new JProperty("version", "1.0.0-" + versionSuffix),
+                    new JProperty("imports", "portable-net451+win8")))));
+
+            File.WriteAllText(projectFile, projectJson.ToString());
+        }
+
+        public void Restore(string project = null)
+        {
+            if (project == null)
+            {
+                project = WorkFolder;
+            }
+            else
+            {
+                project = Path.Combine(WorkFolder, project);
+            }
+
+            // Tests are run in parallel and they try to restore tools concurrently.
+            // This causes issues because the deps json file for a tool is being written from
+            // multiple threads - which results in either sharing violation or corrupted json.
+            lock(_restoreLock)
+            {
+                var restore = ExecuteDotnet($"restore -v Minimal", project);
+                restore.WaitForExit();
+
+                if (restore.ExitCode != 0)
+                {
+                    throw new Exception($"Exit code {restore.ExitCode}");
+                }
+            }
+        }
+
+        private void CreateTestDirectory()
+        {
+            Directory.CreateDirectory(WorkFolder);
+            File.WriteAllText(Path.Combine(WorkFolder, "global.json"), "{}");
+
+            var nugetConfigFilePath = FindNugetConfig();
+
+            var tempNugetConfigFile = Path.Combine(WorkFolder, NugetConfigFileName);
+            File.Copy(nugetConfigFilePath, tempNugetConfigFile);
+        }
+
+        public Process ExecuteDotnet(string arguments, string workDir, IDictionary<string, string> environmentVariables = null)
+        {
+            Console.WriteLine($"Running dotnet {arguments} in {workDir}");
+
+            var psi = new ProcessStartInfo("dotnet", arguments)
+            {
+                UseShellExecute = false,
+                WorkingDirectory = workDir,
+            };
+
+            if (environmentVariables != null)
+            {
+                foreach (var newEnvVar in environmentVariables)
+                {
+                    var varKey = newEnvVar.Key;
+                    var varValue = newEnvVar.Value;
+#if NET451
+                    if (psi.EnvironmentVariables.ContainsKey(varKey))
+                    {
+                        psi.EnvironmentVariables[varKey] = varValue;
+                    }
+                    else
+                    {
+                        psi.EnvironmentVariables.Add(varKey, varValue);
+                    }
+#else
+                    if (psi.Environment.ContainsKey(varKey))
```

Comment:
can't you just do ` psi.Environment[varKey] = varValue;`? Wouldn't it add the entry if the key is missing and update an existing entry?